File: implementations/cpp/tests/benchmark.cpp
#include "pastafari/calendar.hpp"

#include <algorithm>
#include <chrono>
#include <iomanip>
#include <iostream>
#include <vector>

namespace {

template <typename Callable>
double seconds(Callable&& callable) {
    const auto started = std::chrono::steady_clock::now();
    callable();
    const auto finished = std::chrono::steady_clock::now();
    return std::chrono::duration<double>(finished - started).count();
}

}  // namespace

int main() {
    const auto civil = pastafari::GregorianDate::parse("2026-08-06");
    const pastafari::BigInt jdn = pastafari::gregorian_to_jdn(civil);

    std::vector<double> cold;
    for (int sample = 0; sample < 3; ++sample) {
        pastafari::PastafariCalendar calendar;
        cold.push_back(seconds([&] { (void)calendar.convert_jdn(jdn, jdn); }));
    }
    std::sort(cold.begin(), cold.end());

    pastafari::PastafariCalendar calendar;
    (void)calendar.convert_jdn(jdn, jdn);
    constexpr int warm_repetitions = 10'000;
    const double warm = seconds([&] {
        for (int index = 0; index < warm_repetitions; ++index) {
            (void)calendar.convert_jdn(jdn, jdn);
        }
    });
    constexpr int range_days = 365;
    const double sequential = seconds([&] {
        for (int offset = 0; offset < range_days; ++offset) {
            (void)calendar.convert_jdn(jdn, jdn + offset);
        }
    });

    std::cout << std::fixed << std::setprecision(6)
              << "cold median:        " << cold[1] << " s\n"
              << "cached conversion:  "
              << warm / warm_repetitions * 1'000'000.0 << " us/op\n"
              << "365-day sequence:   " << sequential << " s\n";
}

File: implementations/python/benchmark.py
"""Small reproducible benchmark; it is not part of the conformance suite."""

from __future__ import annotations

from statistics import median
from time import perf_counter

from pastafari_calendar import GregorianDate, PastafariCalendar, gregorian_to_jdn


def timed(callable_, repetitions: int = 1) -> float:
    started = perf_counter()
    for _ in range(repetitions):
        callable_()
    return perf_counter() - started


def main() -> None:
    target = GregorianDate.parse("2026-08-06")
    calculation = GregorianDate.parse("2026-08-06")
    calculation_jdn = gregorian_to_jdn(calculation)
    target_jdn = gregorian_to_jdn(target)

    cold_samples = []
    for _ in range(3):
        calendar = PastafariCalendar()
        cold_samples.append(
            timed(lambda: calendar.convert_jdn(calculation_jdn, target_jdn))
        )

    calendar = PastafariCalendar()
    calendar.convert_jdn(calculation_jdn, target_jdn)
    warm_repetitions = 10_000
    warm = timed(
        lambda: calendar.convert_jdn(calculation_jdn, target_jdn),
        warm_repetitions,
    )

    range_days = 365
    sequential = timed(
        lambda: [
            calendar.convert_jdn(calculation_jdn, target_jdn + offset)
            for offset in range(range_days)
        ]
    )

    print(f"cold median:        {median(cold_samples):.6f} s")
    print(f"cached conversion:  {warm / warm_repetitions * 1e6:.3f} us/op")
    print(f"365-day sequence:   {sequential:.6f} s")


if __name__ == "__main__":
    main()
